packages/components/src/lib/controllers/app-controller.js

```diff
@@ -403,83 +394,30 @@ export default class AppController {
 
   async init () {
     if (this.isInitialized) { return 'Already initialized' }
-    // Start loading options as early as possible
-    const optionLoadPromises = this.initOptions(this.options.storageAdapter)
-    const appConfigLoadPromise = this.loadAppConfig({
-      url: this.options.configServiceUrl,
+    // Initialize options
+    await this._stC.init({
+      api: this.api,
+      store: this.store,
+      configServiceUrl: this.options.configServiceUrl,
```

Comment:
This is a tough call. I was following the principle that the App Controller should be a single thing that clients interact with. Clients should not know about existence of internal controllers, the Settings Controller, the UI Controller, etc. So the only way for users to configure anything (even if it is the Settings Controller of which existence they're not aware) is to pass options to the App Controller. It will pass it then along the chain to the specialized controller, if necessary.

Also, different configurations of the App Controller should be possible. In theory, and App Controller might even be used without any specialized controllers at all, not even without the Settings Controller (although I'm not sure how realistic this scenario is). Based on that, the App Controller should keep settings of its own within itself, not trusting it to the Settings Controller. That's why, I though, the App Controller should have its own settings object and its own method (i.e. the `setOptions()`) that will deal with assigning those settings.

@irina060981, @balmas, does it make sense? What do you think?

Reply:
I think that there could be another point of view.
We have an application that couldn't work without options, it has an options tab (that is out of SettingsController), it has options inside lookup box.
If we track options changes inside SettingsController, than it couldn't be excluded without really deep changes in the core architecture and changes of basic principles from my point of view.

And in your case we have duplicate actions for setting options, where appController gets two amount of options and one sets by default and another sends to SettingController from user where they are set again. And for both actions we have completely different methods.

I think that it is useful to have some default settings for the case when we don't have input options, but I suggest to divide options to two states:
   - options as simple input data (from file, from local storage, or for remote source)
   - options as applications objects of Options class

And I agree that all options (as simple input) should go through the APPController as a main input point from the user, but all options as objects of Options class should be handled inside SettingsController.

That is of course only my point of view.

Reply:
I agree it's a tough call. @kirlat I think your reasoning makes sense and I'm fine with it.

Reply:
>And I agree that all options (as simple input) should go through the APPController as a main input point from the user, but all options as objects of Options class should be handled inside SettingsController.


Isn't that essentially what is happening now? The options that are set in the appController are the simple name/value pairs that are configuring the application. I'm missing where this is happening @irina060981 :

>And in your case we have duplicate actions for setting options, where appController gets two amount of options and one sets by default and another sends to SettingController from user where they are set again. And for both actions we have completely different methods.

Reply:
Oh, you are right
we have 
```
this.featureOptionsDefaults = FeatureOptionDefaults
    this.resourceOptionsDefaults = LanguageOptionDefaults
    this.uiOptionsDefaults = UIOptionDefaults
    this.siteOptionsDefaults = SiteOptions
```

and we have merge method for default and loaded
```
static setOptions (options, defaultOptions) {
    let result = {} // eslint-disable-line prefer-const
    for (const [key, defaultValue] of Object.entries(defaultOptions)) {
      // Due to the bug in JS typeof null is `object` and they do not have a `constructor` prop
      // so we have to filter those null values out
      if (typeof defaultValue === 'object' && defaultValue !== null && defaultValue.constructor.name === 'Object') {
        // This is an options group
        const optionsValue = options.hasOwnProperty(key) ? options[key] : {} // eslint-disable-line no-prototype-builtins
        result[key] = this.setOptions(optionsValue, defaultValue)
      } else {
        // This is a primitive type, an array, or other object that is not an options group
        result[key] = options.hasOwnProperty(key) ? options[key] : defaultOptions[key] // eslint-disable-line no-prototype-builtins
      }
    }
    return result
```

I thought that we merge Options objects, but this way we first upload default, then somewhere converts them and use, then get current options for the user, then merge them with default , and again converts to Options objects.

@kirlat , am I right?
it is a little tricky workflow :)

Reply:
`setOptions()` is used to merge options of the App Controller only with their default values. For all other options we use the `initOptions()` method of the Settings Controller.

I agree that the workflow is a little twisted 🙂. I think we can straighten it by loading options defaults directly from within the Settings Controller rather than passing them via the App Controller. I overlooked this simplification and our discussion made me realize that it could be simplified this way.